src/libopensc/card-epass2003.c

```diff
@@ -1330,6 +1330,100 @@ epass2003_sm_wrap_apdu(struct sc_card *card, struct sc_apdu *plain, struct sc_ap
 	return SC_SUCCESS;
 }
 
+static int
+epass2003_check_response_mac_and_sw(struct sc_card *card, struct sc_apdu *sm)
+{
+	unsigned char iv[16];
+	unsigned char *data = NULL, *mac = NULL;
+	size_t blocksize, mac_len;
+	int ret = -1;
+	size_t taglen;
+	const u8 *tag;
+	epass2003_exdata *exdata;
+	unsigned char *in = sm->resp;
+	unsigned char *alt_in;
+	size_t inlen = sm->resplen;
+
+	/* card/ctx/drv_data is already checked by caller */
+	exdata = (epass2003_exdata *)card->drv_data;
+
+	/* compare BER-TLV encoded SW (TAG 0x99) and raw SW */
+	alt_in = in;
+	tag = sc_asn1_find_tag(card->ctx, alt_in, inlen, 0x99, &taglen);
+	if (tag == NULL || taglen != 2) {
+		/*
+		 * It seems that the EPASS2003 firmware has some problem with BER-TLV encoding.
+		 * There seems to be some proprietary fix for the faulty encoding in the
+		 * decrypt_response() function.  Instead of (correct) TLV 87 81 81 [01 ...]
+		 * incorrect TLV 87 81 [01 ...  ] is returned.
```

Comment:
I am confused... both cases are valid BER-TLV structures...
- 87 81 81 [01 ...] is Tag 87, length 129, data starting with 01
- 87 81 [01 ... ] is tag 87, length 1, data is 01

Also, the `if` clause below doesn't seem to be be a workaround for the problem you're describing in the comment.

Reply:
The error appears only and only for the case where the "value" field starts with byte 0x01. (0x01 - padding indicator, and and it is followed by 128 bytes of the answer, therefore the correct value of "L" has the value 129 (encoded 0x81 0x81). Here example (this returns my real epass2003 token):
`````
Incoming APDU (147 bytes):                                                                                                                                        
87 81 01 EA 4E 99 AD 5C 7F 6C 22 23 07 99 7A D2 ....N..\.l"#..z.  
5C 7E 2E 34 5F 13 9B 7A CD C3 86 4E BB D2 7C C9 \~.4_..z...N..|.                                                                                                  
05 9F CF 58 D6 2E 01 F2 60 30 3E F2 0D 4B 89 68 ...X....`0>..K.h  
FD 91 DD 3E 0F D0 EB 4C 0E C0 94 59 E5 2E 7A B6 ...>...L...Y..z.                                                                                                  
E5 15 6F C7 F9 44 0D D4 23 84 CF 37 62 AD 27 6E ..o..D..#..7b.'n  
5A 26 04 82 DD 66 C9 0D 36 F9 37 AF 0C 71 BA 19 Z&...f..6.7..q..                                                                                                  
91 54 60 85 F3 A4 98 2A 78 CF A8 68 3B 6F A0 15 .T`....*x..h;o..  
B0 C0 8A 94 08 D2 23 DD 24 6F FD E0 9C 2C 80 8F ......#.$o...,..                                                                                                  
15 5F 45 99 02 90 00 8E 08 C9 9E 68 58 CE 10 EF ._E........hX...  
A8 90 00                                        ...                                                                                                               
`````

I don't fully understand the problem of why epass2003 returns data incorrectly like this, but the author of the code probably knew why it should be checked this way:

`00b069a08 (Viktor Tarasov       2012-06-04 11:52:50 +0200 1458)         if (0x01 == in[2] && 0x82 != in[1]) {`

Reply:
>  both cases are valid BER-TLV structures...

Both can not be correct BER. It could be BER TLV vs SIMPLE-TLV. The comments indicate it should be BER but in error.